qiita_client/plugin.py

```diff
@@ -0,0 +1,330 @@
+        self.outputs = outputs
+
+    def __call__(self, qclient, server_url, job_id, output_dir):
+        return self.function(qclient, server_url, job_id, output_dir)
+
+
+class QiitaArtifactType(object):
+    """A Qiita artifact type
+
+        Parameters
+        ----------
+        name : str
+            The artifact type name
+        description : str
+            The artifact type description
+        can_be_submitted_to_ebi : bool
+            Whether the artifact type can be submitted to EBI or not
+        can_be_submitted_to_vamps : bool
+            Whether the artifact type can be submitted to VAMPS or not
+        filepath_types : list of (str, bool)
+            The list filepath types that the new artifact type supports, and
+            if they're required or not in an artifact instance of this type"""
+    def __init__(self, name, description, can_be_submitted_to_ebi,
+                 can_be_submitted_to_vamps, filepath_types):
+        self.name = name
+        self.description = description
+        self.ebi = can_be_submitted_to_ebi
+        self.vamps = can_be_submitted_to_vamps
+        self.fp_types = filepath_types
+
+
+class BaseQiitaPlugin(object):
+    def __init__(self, name, version, description, publications=None):
+        self.name = name
+        self.version = version
+        self.description = description
+        self.publications = dumps(publications) if publications else ""
+
+        # Will hold the different commands
+        self.task_dict = {}
+
+        # The configuration file
+        conf_dir = environ.get(
+            'QIITA_PLUGINS_DIR', join(expanduser('~'), '.qiita_plugins'))
+        self.conf_fp = join(conf_dir, "%s_%s.conf" % (self.name, self.version))
+
+    def generate_config(self, env_script, start_script, server_cert=None):
+        """Generates the plugin configuration file"""
+        sr = SystemRandom()
+        chars = ascii_letters + digits
+        client_id = ''.join(sr.choice(chars) for i in range(50))
+        client_secret = ''.join(sr.choice(chars) for i in range(255))
+
+        server_cert = server_cert if server_cert else ""
+
+        with open(self.conf_fp, 'w') as f:
+            f.write(CONF_TEMPLATE % (self.name, self.version, self.description,
+                                     env_script, start_script,
+                                     self._plugin_type, self.publications,
+                                     server_cert, client_id, client_secret))
+
+    def _register_command(self, command):
+        """Registers a command in the plugin
+
+        Parameters
+        ----------
+        command: QiitaCommand
+            The command to be added to the plugin
+        """
+        self.task_dict[command.name] = command
+
+    def _install(self, qclient):
+        """Installs the plugin in Qiita"""
+        # Get the command information from qiita
+        info = qclient.get('/qiita_db/plugins/%s/%s/'
```

Comment:
Yes, so when Qiita starts the server, it will load the plugin configuration files. The reason behind this is that in order to communicate the plugin (client) and Qiita (server) they need to both know the credentials. So the config file is generating those credentials, so in order to actually add a plugin to the system you need access to the FS to be able to add this configuration file. However, the configuration file only allows to "register" the plugin itself, but not the commands. The commands get added dynamically using the REST api, which is what is done here.

Is a bit complex, but I think this is the best balance between security and flexibility. This flexibility in the commands is needed mainly for QIIME 2, which is a system that works with plugins itself, so it is possible that there are commands that are no longer available, while new ones may appear.

Regarding the 200, the QiitaClient object (qclient) already performs those checks internally and it will raise a useful error in case that the return code is different from 200. The object returned from the `get` and `post` is actually a dictionary, which contains the decoded JSON from the request body.